Add Live Wallpaper sample, fixes #298

koral-- opened the pull request.

AndroidDeveloperLB commented:
@koral-- It shows me the exact same thing:

![image](https://user-images.githubusercontent.com/5357526/40994619-63b9cc0e-6905-11e8-8aab-7ffcb3ecf11e.png)

But using the zip option works.
Which scaling is used there exactly? 
What are the rules of how it's used?

koral-- commented:
After cloning you can checkout the `live_wallpaper` branch e.g. using `git checkout live_wallpaper` or use equivalent option from GUI e.g. in Android Studio: 
![image](https://user-images.githubusercontent.com/3340954/40995667-fdeb3930-68ff-11e8-9a04-a942029a51bd.png)

Currently the GIF is displayed in center of the surface. In terms of ImageView scale types it is center_inside.

AndroidDeveloperLB commented:
How do I choose it in TurtoiseGit? I get this:

![image](https://user-images.githubusercontent.com/5357526/40995864-0ac0fc54-6909-11e8-90e3-ec3e18c68a2d.png)

Also, about center-inside, just to verify: this means that it will try to use the original size, keep aspect ratio, and if it's too large compared to screen size, it will auto-fit (and still keep aspect ratio of course) ?

koral-- commented:
It seems that you need to check `Branch` checkbox and type `live_wallpaper` on the right of it.

> Also, about center-inside, just to verify: this means that it will try to use the original size, keep aspect ratio, and if it's too large compared to screen size, it will auto-fit (and still keep aspect ratio of course) ?

Original aspect ratio will be kept. However, case when GIF is larger than screen size is not handled. GIF used in this sample is smaller than any screen size. If it is replaced with larger one it will exceed screen size (parts outside won't be visible).

AndroidDeveloperLB commented:
How can I know [here ](https://github.com/koral--/android-gif-drawable/tree/7fc31357a10df700d151323a4ce3fd22fb838dc8)that I should write "live_wallpaper" ? 

About the scaling, so it's not exactly like center-inside... 
Center-inside doesn't let this happen.

koral-- commented:
You are right, I'll fix that.

AndroidDeveloperLB commented:
Trying to clone from TurtoisGit seems to have an issue (very very long cloning).
Using the Zip file, I can see this build issue:

![image](https://user-images.githubusercontent.com/5357526/41745324-3f25145a-75af-11e8-8bcd-31acd10e2f1b.png)

AndroidDeveloperLB commented:
ok I've removed some stuff from gradle, of "shouldBuildArmeabi" and maybe some more, and now it got built and I ran it, but even though the preview of the wallpaper was fine, this is what I get on the launcher:

![image](https://user-images.githubusercontent.com/5357526/41759646-28e76ea6-75f8-11e8-812c-fa734a2b1a65.png)

There is also a memory leak notification there.

koral-- commented:
> Trying to clone from TurtoisGit seems to have an issue (very very long cloning).

It seems to be an issue on your/yours ISP side.

>  I can see this build issue:

Fixed.

> this is what I get on the launcher:

Fixed.

> There is also a memory leak notification there.

Fixed those originating from sample project code.

AndroidDeveloperLB commented:
I don't know about the cloning. I've tried now at the office too (different PC), and got this issue here too.
It cloned for a very long time, said it finished, but when I got into the folder, there's nothing there:

![image](https://user-images.githubusercontent.com/5357526/41973845-dd825342-7a1e-11e8-99ea-9c9fdf424473.png)


About the new sample (had to download using zip), it still had a memory leak notification while swiping between the pages of the Viewpager.

About the live wallpaper, it still has serious issues, especially when viewing in landscape. It also has some memory leak notification.
Attached video:

[device-2018-06-27-155831.zip](https://github.com/koral--/android-gif-drawable/files/2141266/device-2018-06-27-155831.zip)

AndroidDeveloperLB commented:
Oh, and about the cloning, I've found the issue. It's supposed to be this way:

![image](https://user-images.githubusercontent.com/5357526/42031020-7ab5d2f6-7add-11e8-89fb-2df6876f338e.png)

AndroidDeveloperLB commented:
Got a crash while swiping between pages in ViewPager. Log doesn't seem to make sense, but I write it anyway:

2018-07-02 09:48:30.915 607-1915/? W/SurfaceFlinger: Attempting to destroy on removed layer: 4fd56a6 Application Error: pl.droidsonroids.gif.sample#0
    
    --------- beginning of system
2018-07-02 09:48:33.179 7504-7504/pl.droidsonroids.gif.sample W/oids.gif.sampl: Accessing hidden field Ljava/io/FileDescriptor;->descriptor:I (light greylist, JNI)
2018-07-02 09:48:33.230 7504-7504/pl.droidsonroids.gif.sample W/oids.gif.sampl: Accessing hidden field Landroid/view/View;->mAccessibilityDelegate:Landroid/view/View$AccessibilityDelegate; (light greylist, reflection)
2018-07-02 09:48:33.558 607-667/? W/SurfaceFlinger: Attempting to set client state on removed layer: Splash Screen pl.droidsonroids.gif.sample#0
2018-07-02 09:48:33.558 607-667/? W/SurfaceFlinger: Attempting to destroy on removed layer: Splash Screen pl.droidsonroids.gif.sample#0
2018-07-02 09:48:35.148 7504-7523/pl.droidsonroids.gif.sample W/Adreno-GSL: <gsl_ldd_control:548>: ioctl fd 55 code 0xc040094a (IOCTL_KGSL_GPU_COMMAND) failed: errno 22 Invalid argument
2018-07-02 09:48:35.164 7504-7523/pl.droidsonroids.gif.sample W/Adreno-GSL: <gsl_syncobj_dup_fd:265>: 0x7d005d70e0 dup(84) error 9 Bad file descriptor
2018-07-02 09:48:35.182 7504-7523/pl.droidsonroids.gif.sample W/Adreno-GSL: <gsl_syncobj_dup_fd:265>: 0x7d007fe380 dup(84) error 9 Bad file descriptor
2018-07-02 09:48:35.199 7504-7523/pl.droidsonroids.gif.sample W/Adreno-GSL: <gsl_syncobj_dup_fd:265>: 0x7d007fe480 dup(84) error 9 Bad file descriptor
2018-07-02 09:48:35.215 7504-7523/pl.droidsonroids.gif.sample W/Adreno-GSL: <gsl_syncobj_dup_fd:265>: 0x7d005d70e0 dup(84) error 9 Bad file descriptor
2018-07-02 09:48:35.233 7504-7523/pl.droidsonroids.gif.sample W/Adreno-GSL: <gsl_syncobj_dup_fd:265>: 0x7d007fe380 dup(84) error 9 Bad file descriptor
2018-07-02 09:48:35.248 7504-7523/pl.droidsonroids.gif.sample W/Adreno-GSL: <gsl_syncobj_dup_fd:265>: 0x7d007fe480 dup(84) error 9 Bad file descriptor
2018-07-02 09:48:35.266 7504-7523/pl.droidsonroids.gif.sample W/Adreno-GSL: <gsl_syncobj_dup_fd:265>: 0x7d007fe380 dup(84) error 9 Bad file descriptor
2018-07-02 09:48:37.289 7504-7523/pl.droidsonroids.gif.sample E/GLConsumer: [SurfaceTexture-0-7504-0] detachFromContext: abandoned GLConsumer
2018-07-02 09:48:37.289 7504-7523/pl.droidsonroids.gif.sample E/OpenGLRenderer: Failed to detach SurfaceTexture from context -19
2018-07-02 09:48:38.257 7504-7523/pl.droidsonroids.gif.sample E/GLConsumer: [SurfaceTexture-0-7504-1] detachFromContext: abandoned GLConsumer
2018-07-02 09:48:38.257 7504-7523/pl.droidsonroids.gif.sample E/OpenGLRenderer: Failed to detach SurfaceTexture from context -19
2018-07-02 09:48:38.528 7504-7523/pl.droidsonroids.gif.sample E/GLConsumer: [SurfaceTexture-0-7504-2] detachFromContext: abandoned GLConsumer
2018-07-02 09:48:38.528 7504-7523/pl.droidsonroids.gif.sample E/OpenGLRenderer: Failed to detach SurfaceTexture from context -19
2018-07-02 09:48:40.529 7504-7523/pl.droidsonroids.gif.sample W/Adreno-GSL: <gsl_ldd_control:548>: ioctl fd 55 code 0xc040094a (IOCTL_KGSL_GPU_COMMAND) failed: errno 22 Invalid argument
    
    --------- beginning of crash
2018-07-02 09:48:40.534 7504-7542/pl.droidsonroids.gif.sample E/AndroidRuntime: FATAL EXCEPTION: GifIOThread
    Process: pl.droidsonroids.gif.sample, PID: 7504
    java.lang.RuntimeException: Could not read from eventfd Success
2018-07-02 09:48:40.538 1145-13657/? W/ActivityManager:   Force finishing activity pl.droidsonroids.gif.sample/.MainActivity
2018-07-02 09:48:40.695 607-626/? W/SurfaceFlinger: Attempting to destroy on removed layer: AppWindowToken{95e8b77 token=Token{3c9e976 ActivityRecord{eb61411 u0 pl.droidsonroids.gif.sample/.MainActivity t25621}}}#0
2018-07-02 09:48:40.865 1145-1278/? E/LoadedApk: Unable to instantiate appComponentFactory
    java.lang.ClassNotFoundException: Didn't find class "android.support.v4.app.CoreComponentFactory" on path: DexPathList[[],nativeLibraryDirectories=[/data/app/lb.com.giflivewallpaper-QCQafAr7WgOBp_0M9hhD7Q==/lib/arm64, /system/lib64]]
        at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:126)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:379)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:312)
        at android.app.LoadedApk.createAppFactory(LoadedApk.java:226)
        at android.app.LoadedApk.createOrUpdateClassLoaderLocked(LoadedApk.java:703)
        at android.app.LoadedApk.getClassLoader(LoadedApk.java:772)
        at android.app.ContextImpl.getClassLoader(ContextImpl.java:332)
        at android.view.LayoutInflater.createView(LayoutInflater.java:606)
        at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:790)
        at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:730)
        at android.view.LayoutInflater.inflate(LayoutInflater.java:492)
        at android.view.LayoutInflater.inflate(LayoutInflater.java:423)
        at android.view.LayoutInflater.inflate(LayoutInflater.java:374)
        at com.android.internal.policy.DecorView.onResourcesLoaded(DecorView.java:1888)
        at com.android.internal.policy.PhoneWindow.generateLayout(PhoneWindow.java:2594)
        at com.android.internal.policy.PhoneWindow.installDecor(PhoneWindow.java:2667)
        at com.android.internal.policy.PhoneWindow.getDecorView(PhoneWindow.java:2066)
        at com.android.server.policy.PhoneWindowManager.addSplashScreen(PhoneWindowManager.java:3173)
        at com.android.server.wm.SplashScreenStartingData.createStartingSurface(SplashScreenStartingData.java:56)
        at com.android.server.wm.AppWindowContainerController$1.run(AppWindowContainerController.java:151)
        at android.os.Handler.handleCallback(Handler.java:873)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:193)
        at android.os.HandlerThread.run(HandlerThread.java:65)
        at com.android.server.ServiceThread.run(ServiceThread.java:44)
2018-07-02 09:48:41.032 607-667/? W/SurfaceFlinger: Attempting to destroy on removed layer: 74f5c0e Application Error: pl.droidsonroids.gif.sample#0
2018-07-02 09:48:41.172 7550-7568/? W/ProcStatsManager: No process lb.com.giflivewallpaper/12016 for service lb.com.giflivewallpaper.GIFWallpaperService
2018-07-02 09:48:41.186 7550-7568/? W/ProcStatsManager: No process lb.com.giflivewallpaper/12016 for service lb.com.giflivewallpaper.GIFWallpaperService

Also, still memory leak.

Live wallpaper itself seems to work well, but how can I configure how to scale it? 
Which type of scaling does it use?

AndroidDeveloperLB commented:
It still crashes upon switching between pages in the ViewPager.
Is it a problem in the sample, or the library itself?

AndroidDeveloperLB commented:
And how can I choose which scaling to use, in the sample?

koral-- commented:
Sorry for my poor responsiveness.

> It still crashes upon switching between pages in the ViewPager.
Is it a problem in the sample, or the library itself?

Hard to say without proper error message stacktrace. Does it also happen on `release` build variant? If not then it seems to be caused by Leak Canary. I've noticed that it sometimes fails to check for leaks and causes process abort. Leak canary is disabled on release version of sample.
 
I've only spotted this bug in your log: `Could not read from eventfd Success`. It should be fixed now.

> And how can I choose which scaling to use, in the sample?

I've changed this now a little bit.
Here is the logic: https://github.com/koral--/android-gif-drawable/pull/550/commits/80184f6ca606bf82fe995600bd16b1c80da89da1#diff-06731a942b8945151a71752870b04055R75

It uses viewport to achieve different scaling modes. Viewport is like virtual screen on which we draw the GIF frames. Content from GIF fills entire viewport. However, viewport not necessarily fills entire real screen.
In case of FIT_CENTER viewport has the same aspect ratio as GIF and one of the dimensions equal to corresponding one from GIF. It is also moved to the center of real screen. Point (0;0) is on lower left corner.
In case of FIT_XY viewport has the same dimensions and position as real screen.

koral-- commented:
> In "release" we have the live wallpaper?

Yes, it should work.

> Also, can you please put other scaling types? center-crop, center-inside ?

I'll do that in another PR.

> When you write about "Point (0;0) " what does it mean about the live wallpaper? That the content will be on the lower left corner, and not the center?

If you set (x, y) to (0;0) then lower left corner of the wallpaper will be exactly  at the lower left corner of the screen. If for example scaled wallpaper needs to be moved 200 px up to be centered then you should use (0;200).

AndroidDeveloperLB commented:
I don't see "release" branch :

![image](https://user-images.githubusercontent.com/5357526/42344095-f12a0ce2-80a3-11e8-955d-3e786bac8de1.png)

About the scaling, do you know perhaps how to do it for other types of files? WebP ? Animated HEIF? Video files? Is it possible to have a single solution for them all?
I've asked about these here:
https://stackoverflow.com/q/49944661/878126
https://stackoverflow.com/q/50091878/878126
I will be glad to grant the bounties, of course, if you show how exactly to do those.

I don't understand about the 0,0 point. Do you mean that for each frame you need to move the content, because it's supposed to be centered, but the frames can have different sizes?

koral-- commented:
> I don't see "release" branch :

I meant `release` build variant, not branch. It can be selected in Android Studio:
![image](https://user-images.githubusercontent.com/3340954/42352269-9bea8ade-80b9-11e8-93e5-5fe2cffc0718.png)

> About the scaling, do you know perhaps how to do it for other types of files? WebP ? Animated HEIF? Video files? Is it possible to have a single solution for them all?

If you replace GifTexImage2D in sample with some other source which can render to GL texture then it should work for other file types. For videos you can use for example Exoplayer. Not sure if there is a single library which can handle all file types you mentioned.

> I don't understand about the 0,0 point. 

I meant that the starting point is located on the lower left corner unlike e.g. Android View coordinate system when it is on upper left corner.

AndroidDeveloperLB commented:
Regarding release, if you mean release build of the "live_wallpaper" branch, for some reason I can't build it. Not sure which parts of the gradle files I need to remove (I tried).

Regarding the scaling, still would love to see if you succeed it on other cases.

Regarding the 0,0 point, have you overcome it though?

AndroidDeveloperLB commented:
If this is supposed to fix the leaks or crashes, it didn't work.

AndroidDeveloperLB commented:
You mean to add this:

    debugImplementation "com.squareup.leakcanary:leakcanary-android-no-op:$versions.leakCanary"

instead of this:

    debugImplementation "com.squareup.leakcanary:leakcanary-android:$versions.leakCanary"

If so, after some swiping, then changing orientation, and then some more swiping, I've found that it got ANR (and doesn't let me swipe at all, even if I tell it to wait) :
![image](https://user-images.githubusercontent.com/5357526/42435489-725975b0-835f-11e8-9d99-67324c4d3475.png)

About the scaling, there is a nice library that shows a lot more than what Google has added for ImageView, [here](https://github.com/yqritc/Android-ScalableVideoView). Maybe it could help you?

koral-- commented:
I could not repro that ANR on Pixel 2 with P DP4.
Could you paste ANR trace?

AndroidDeveloperLB commented:
It took me some time. 
Now that I've tested it, I couldn't reproduce it, but I've noticed other things. I wanted to take a video for the ANR, but I failed, so this is what I've found now: 
[device-2018-07-09-150816.zip](https://github.com/koral--/android-gif-drawable/files/2176039/device-2018-07-09-150816.zip)

- GifTextuView and HTTP - sometimes show flag in much larger size
- GifDecoder - doesn't change colors so often (between blue and yellow backgrounds), and not in same frequency.

After second try, I got the ANR. Attached log file.
[log.txt](https://github.com/koral--/android-gif-drawable/files/2176065/log.txt)
